internal/recon/md5.go

```diff
@@ -0,0 +1,165 @@
+		notMatched: promhelper.NewGaugeTypedDesc(
+			"swift_cluster_md5_not_matched",
+			"Not matched hosts for md5sum(s) reported by the swift-recon tool.", []string{"storage_ip", "kind"}),
+	}
+}
+
+// describeMetrics implements the task interface.
+func (t *md5Task) describeMetrics(ch chan<- *prometheus.Desc) {
+	t.all.Describe(ch)
+	t.errors.Describe(ch)
+	t.matched.Describe(ch)
+	t.notMatched.Describe(ch)
+}
+
+// md5OutputBlockRx extracts the entire output block for a specific kind from
+// the aggregate md5 recon output.
+//
+// Match group ref:
+//  <1: kind> <2: output block>
+//
+// Example of an output block for ring:
+//   [<time-stamp>] Checking ring md5sums
+//   -> On disk object.ring.gz md5sum: 123456
+//   -> http://10.0.0.1:6000/recon/ringmd5: <urlopen error [Errno 111] ECONNREFUSED>
+//   -> http://10.0.0.2:6000/recon/ringmd5: {'/path/to/account.ring.gz': '123456', '/path/to/container.ring.gz': '123456', '/path/to/object.ring.gz': '123456'}
+//   -> http://10.0.0.2:6000/recon/ringmd5 matches.
+// 1/2 hosts matched, 1 error[s] while checking hosts.
+var md5OutputBlockRx = regexp.MustCompile(
+	`(?m)^.* Checking ([\.a-zA-Z0-9_]+) md5sums?\s*((?:(?:->|!!)\s\S*\s.*\s*)*)\s*[0-9]+/[0-9]+ hosts matched, [0-9]+ error.*$`)
+
+// collectMetrics implements the task interface.
+func (t *md5Task) collectMetrics(ch chan<- prometheus.Metric, exitCodeTypedDesc *promhelper.TypedDesc) {
+	exitCode := 0
+	cmdArgs := []string{fmt.Sprintf("--timeout=%d", t.hostTimeout), "--md5"}
+	defer func() {
+		ch <- exitCodeTypedDesc.MustNewConstMetric(float64(exitCode), cmdArgsToStr(cmdArgs))
+	}()
+
+	var matchList [][][]byte
+	out, err := runCommandWithTimeout(t.ctxTimeout, t.pathToReconExecutable, cmdArgs...)
+	if err == nil {
+		matchList = md5OutputBlockRx.FindAllSubmatch(out, -1)
+		if len(matchList) == 0 {
+			err = fmt.Errorf("command did not return any usable output:\n%s", string(out))
+		}
+	}
+	if err != nil {
+		exitCode = 1
+		logg.Error("swift recon: %s: %s", cmdArgsToStr(cmdArgs), err.Error())
+		return
+	}
+
+	for _, match := range matchList {
+		kind := string(match[1])
+		outputBlock := match[2]
+		outputPerHost, err := splitOutputPerHost(outputBlock, cmdArgs)
+		if err != nil {
+			exitCode = 1
+			logg.Error("swift recon: %s: %s: output follows:\n%s", cmdArgsToStr(cmdArgs), err.Error(), string(outputBlock))
+			continue // to next output block
+		}
+
+		// allHostsMatch and errHosts are used to submit zero value metrics so
+		// that we can use aggregation rules in case that MD5 of all hosts for
+		// a specific kind matches.
+		allHostsMatch := true
+		// We use a map for errHosts to maintain a record of unique hostnames
+		// that had errors since there could be multiple errors per host.
+		errHosts := make(map[string]bool)
+		var all float64
+		for hostname, dataBytes := range outputPerHost {
+			// Host output can be in the following formats:
+			//   1. {'/path/to/<server-type>.ring.gz': '<md5-sum>', ...}
```

Comment:
I was confused here because this is checked with `json.Valid()`, but JSON does not allow single quotes. The example should reflect what format is actually going into here:

```suggestion
			//   1. {"/path/to/<server-type>.ring.gz": "<md5-sum>", ...}
```